### ISSUE TITLE
Loading of XML files can fail resulting in VCard import failures

### ISSUE THREAD

**Issue by AccountToPostOwncloudIssue** [owncloud/contacts#1132]
On my Owncloud installation, importing contacts stored in a VCard file failed. it irregularly actually worked, but most of the times it failed. It just finished with the message 0 contacts imported. 

I tracked the problem down to the code in `importmanager.php` that reads the list of available importers by parsing their XML files. I got this error message by libxml:
```failed to load external entity "/var/www/owncloud/apps/contacts/formats/import_csv_outlook_connector.xml"```  

When I added a call to `libxml_disable_entity_loader(false)` before the parsing of the XML file, it worked for me (but I didn't test it very much), so I think that maybe somewhere else in the application loading of external entities is disabled causing problems here.

Note: I am still using owncloud 8.1.5.2, so it could be that this problem is actually already resolved in the newest version. In that case, sorry for reporting it unnecessarily.

**Comment by babelouest** [comment]
That's odd, you seem to have a problem with external entity that are loaded by `simpleXML`, but why ?

Have you checked your apache/php installation to be clean ?

For now I don't think I'll add `libxml_disable_entity_loader(false)` like you did because it's said to be non thread-safe, and I don't want owncloud to cause other problems in people's configuration.

Is there any other web application that could cause a configuration conflict with owncloud ?

**Comment by AccountToPostOwncloudIssue** [comment]
I did a few more experiments, and the behaviour is really strange. I created this simple test file and put it on the server:

```
<?php
        $xml = simplexml_load_file('/var/www/owncloud/apps/contacts/formats/import_vcard_gmail_connector.xml');
        echo('Parsed: ' . print_r($xml, true));

        $loadEntities = libxml_disable_entity_loader(true);
        libxml_disable_entity_loader($loadEntities);
        echo('Old Setting: ' . print_r($loadEntities, true));
?>
```

If I reload this file repeatedly, it sometimes parses the XML (and returns nothing as `$loadedEntities`) and sometimes it does not load the file, but instead give this warning:
```
Warning: simplexml_load_file(): I/O warning : failed to load external entity "/var/www/owncloud/apps/contacts/formats/import_vcard_gmail_connector.xml" in /var/www/owncloud/test.php on line 3
Parsed:
```
and returns 1 as last setting.

I haven't found any documentation on what the default value for this setting is (http://php.net/manual/en/function.libxml-disable-entity-loader.php doesn't say anything about this). I also didn't find a setting in the php.ini. Could it be, that the state of this setting is unreliable? If the function is not threadsafe, can multiple running apache threads influence each other? Or could it be that the state depends on the last script that has been executed in the same server process before the current one is executed? When I call php on the script above from the command line the behaviour is actually deterministic (it always loads the xml file).

In https://www.mediawiki.org/wiki/XML_External_Entity_Processing they recommend for security reasons to always set reading of external entities to false and to read the xml into a string and parse that string.

**Comment by AccountToPostOwncloudIssue** [comment]
> Is there any other web application that could cause a configuration conflict with owncloud ?

The server is also running roundcube and tiny-tiny-rss.

**Comment by AccountToPostOwncloudIssue** [comment]
Yes, it seems that the setting is preserved over multiple script executions. When I set it to a certain state at the end of the test script, after a few reloads, it will always have this value at the beginning of the script. So after a few reloads it seems that all server-threads were changed to the new value. When I then change the value in the php and reload again, they will switch back after more reloads. 

In contrast, the script above changed it to two different values. So it seems that, either due to multithreading or due to interruption of the execution via F5, this caused the value to toggle.

Unless there is still some configuration option I am overlooking, explicitly setting the value seems to be the only way to be safe. However, if it is a multi-threading problem not even that may suffice. There are already several places in the owncloud code where the function `libxml_disable_entity_loader` is called (both with true and false as parameter). So even a server running just owncloud wouldn't be safe from multi-threading problems.
